### PR TITLE
[NPU] Reuse prefill of acc lib for pipeline

### PR DESCRIPTION
**Pull request by rnwang04**
### 1. Why the change?

Accelerate prefill of NPU pipeline.
Work with https://github.com/intel-analytics/llm.cpp/pull/598

### 2. User API changes

See example.

### 3. Summary of the change 

- Reuse prefill of acc lib for pipeline
- Modify `generate` function for corresponding adaptation
- Update embedding weight from input to const
- Update examples
- change `max-output-len` to `max-context-len`

### 4. How to test?
- [x] Unit test: Please manually trigger the PR Validation [here](https://github.com/intel-analytics/ipex-llm-workflow/actions/workflows/llm-PR-validation.yml) by inputting the PR number (e.g., `1234`). And paste your action link here once it has been successfully finished.
- [x] Application test

### 5. Demo output
image

### CODE REVIEW

**Comment by rnwang04** [comment]
PR validation is passed: https://github.com/intel-analytics/ipex-llm-workflow/actions/runs/11545625174/job/32132875794

File: python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py [hunk 1]
```diff
@@ -182,8 +219,29 @@ def update_names_of_IR_and_export_blob(model, model_name, dir):
...
+            max_prompt_len=max_prompt_len,
+            transpose_value_cache=transpose_value_cache,
+        )
+        llama_model_forward = gen_llama_fused_model_forward(
+            prefill_runner=prefill_runner, decode_runner=None
+        )
+        convert_forward(model, LlamaModel, llama_model_forward)
+        from transformers.models.llama.modeling_llama import LlamaForCausalLM
+        from ipex_llm.transformers.npu_models.llama_mp import llama2_casullm_forward
+        convert_forward(model, LlamaForCausalLM, llama2_casullm_forward)
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py:243]: Refactor this in a common method to be called by both pipeline and accelerate models.

Comment (rnwang04) [reply python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py]: Have refactored this part as you said. Would you mind taking a look again ?

**Comment by rnwang04** [comment]
Performance validation is updated here : https://github.com/analytics-zoo/nano/issues/1687#issuecomment-2440336530

File: python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py [hunk 2]
```diff
@@ -63,85 +63,121 @@ def generate(
...
+                output_pipe = open(out_pipe_path, "rb")
+            except:
+                print('Waiting for output pipe')
+                time.sleep(1)
+            else:
+                break
+
+        time_start = time.perf_counter()
+
+        bdata = str.encode(str(temp_dir))
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py:148]: check length of `str(temp_dir)` here (<2000?)

Comment (rnwang04) [reply python/llm/src/ipex_llm/transformers/npu_pipeline_model/convert_pipeline.py:148]: Added check here.

File: python/llm/src/ipex_llm/transformers/npu_model.py [hunk 3]
```diff
@@ -259,6 +259,7 @@ def optimize_npu_model(cls, *args, **kwargs):
                 import convert_llm
             convert_llm(llm,
                         kv_len=max_output_len,
+                        max_prompt_len=max_prompt_len,
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_model.py:262]: `max_prompt_len` should be smaller than 4096 in the current implementation?

Comment (rnwang04) [reply python/llm/src/ipex_llm/transformers/npu_model.py:262]: I think in the current implementation, `max_prompt_len` should be smaller than `kv_len` (like 1023) ?

Comment (jason-dai) [reply python/llm/src/ipex_llm/transformers/npu_model.py:262]: In the cpp implementation, the prompt buffer is set to 4096; and is there a constraint of `kv_len` or `max_output_len`?

Comment (rnwang04) [reply python/llm/src/ipex_llm/transformers/npu_model.py:262]: In the latest implementation, prompt buffer is not used now.
There is no constraint of `kv_len` or` max_output_len`, but has constraint for new tokens now as limitation of `Init_pipe()`.
Added a check for this first.

**Review by jason-dai — APPROVED** [review]
LGTM

**Pull request merged**